File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(records):
    dic = {}
    answer = []
    for record in records:
        # command = Enter, Leave, Change
        if record.startswith("Leave"):
            command, uid = record.split()
            answer.append([command, uid])
        elif record.startswith("Enter"):
            command, uid, name = record.split()
            dic[uid] = name
            answer.append([command, uid])
        else:
            command, uid, name = record.split()
            dic[uid] = name

    result = []
    for (command, uid) in answer:
        if command == "Enter":
            result.append(f"{dic[uid]}님이 들어왔습니다.")
        else:
            result.append(f"{dic[uid]}님이 나갔습니다.")

    return result

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        Map<String, String> map = new HashMap<>();
        
        List<List<String>> result = new ArrayList<>();
        
        for (String r : record) {
            if (r.startsWith("Enter")) {
                String[] split = r.split(" ");
                map.put(split[1], split[2]);
                result.add(List.of("Enter", split[1]));
            } else if (r.startsWith("Change")) {
                String[] split = r.split(" ");
                map.put(split[1], split[2]);
            } else {
                String[] split = r.split(" ");               
                result.add(List.of("Leave", split[1]));
            }
        }
        
        String[] arr = new String[result.size()];
        for (int i=0; i<result.size(); i++) {
            String id = map.get(result.get(i).get(1));
            if (result.get(i).get(0).equals("Enter")) {
                arr[i] = String.format("%s님이 들어왔습니다.", id);
            } else {
                arr[i] = String.format("%s님이 나갔습니다.", id);                               
            }
        }
        
        return arr;
    }
}